File: python_concept/search_rag.py
#!/usr/bin/env python3
"""search_rag.py — busca na base RAG (a fase ONLINE).

Consome a base gerada pelo embed_gen.py (ex: sda-tokenized.json) e responde queries.
Como a base e' AUTO-SUFICIENTE (vocab + idf + vetores inline), a busca NAO reprocessa
o corpus — so carrega o JSON e pontua.

Pipeline (o mesmo do RAG de producao):
  estagio 1 — RECALL : cosseno da query com cada chunk (usa o vec esparso + norm da base)
  estagio 2 — RERANK : matched filter (contiguidade) + span (proximidade) sobre o texto
  -> top-k com snippet keyword-in-context (palavras da query em DESTAQUE).

Uso (a base JSON e' obrigatoria — sempre explicita, nunca inferida):
  python3 search_rag.py sda-tokenized.json "Frodo Bolseiro"      # top-5
  python3 search_rag.py sda-tokenized.json "anel" "montanha" -k 3
  python3 search_rag.py sda-tokenized.json "Gandalf" --no-rerank # so recall cosseno
  python3 search_rag.py sda-tokenized.json -i                    # modo interativo (REPL)
"""
import os, sys, json, time, unicodedata, argparse

# sylkit mora na raiz do projeto (ao lado deste arquivo) — import direto
from sylkit import syllabify, normalize, syllable_seq, WORD, tfidf, cosine

PROX_SCALE = 8.0   # rr = mf / (1 + span/PROX_SCALE): penaliza palavras distantes


# ---------------------------------------------------------------------------
# estagio 2 — rerank: contiguidade (matched filter) + proximidade (span)
# ---------------------------------------------------------------------------
def _best_positions(qs, seq):
    """Melhor fracao de casamento contiguo de qs em seq + posicoes onde ocorre."""
    k = len(qs)
    if k == 0 or len(seq) < k:
        return 0.0, [0]
    best, pos = -1, []
    for p in range(len(seq) - k + 1):
        match = sum(1 for j in range(k) if seq[p + j] == qs[j])
        if match > best:
            best, pos = match, [p]
        elif match == best:
            pos.append(p)
    return best / k, pos


def _min_span(lists):
    """Menor janela que cobre uma posicao de CADA lista (smallest range)."""
    import heapq
    heap = [(lst[0], i, 0) for i, lst in enumerate(lists)]
    heapq.heapify(heap)
    cur_max = max(lst[0] for lst in lists)
    best = cur_max - heap[0][0]
    while True:
        mn, i, j = heapq.heappop(heap)
        if cur_max - mn < best:
            best = cur_max - mn
        if j + 1 == len(lists[i]):
            return best
        nxt = lists[i][j + 1]
        cur_max = max(cur_max, nxt)
        heapq.heappush(heap, (nxt, i, j + 1))


def rerank_score(query, seq):
    """Combina contiguidade (mf) e proximidade (span) das palavras do query."""
    words = [[normalize(s) for s in syllabify(w) if normalize(s)]
             for w in WORD.findall(query.lower())]
    words = [qs for qs in words if qs]
    if not words:
        return 0.0, 0
    fracs, lists = [], []
    for qs in words:
        frac, pos = _best_positions(qs, seq)
        fracs.append(frac); lists.append(pos)
    mf = sum(fracs) / len(fracs)
    span = _min_span(lists) if len(lists) > 1 else 0
    return mf, span


def _strip(s):
    s = unicodedata.normalize("NFD", s.lower())
    return "".join(c for c in s if unicodedata.category(c) != "Mn")


def snippet(text, query, width=140):
    """Trecho centrado na 1a palavra do query, com as palavras em «destaque»."""
    flat = " ".join(text.split())
    flat_na = _strip(flat)
    qwords = WORD.findall(query.lower())
    pos = -1
    for w in qwords:
        pos = flat_na.find(_strip(w))
        if pos >= 0:
            break
    if pos < 0:
        out = (flat[:width] + "…") if len(flat) > width else flat
    else:
        a, b = max(0, pos - 35), min(len(flat), pos - 35 + width)
        out = ("…" if a else "") + flat[a:b].strip() + ("…" if b < len(flat) else "")
    # destaca as palavras da query (case-insensitive, sem acento)
    out_na = _strip(out)
    marks = []
    for w in qwords:
        wn, start = _strip(w), 0
        while True:
            i = out_na.find(wn, start)
            if i < 0:
                break
            marks.append((i, i + len(wn))); start = i + len(wn)
    for a, b in sorted(set(marks), reverse=True):
        out = out[:a] + "«" + out[a:b] + "»" + out[b:]
    return out


# ---------------------------------------------------------------------------
# a base RAG carregada em memoria
# ---------------------------------------------------------------------------
class RagBase:
    def __init__(self, path):
        with open(path, encoding="utf-8") as f:
            b = json.load(f)
        self.meta = b["meta"]
        self.vocab = self.meta["vocab"]
        self.index = {t: i for i, t in enumerate(self.vocab)}
        self.idf = {int(k): v for k, v in b["idf"].items()}
        # chunks: vec esparso (chaves str -> int), indexado por id
        self.chunks = {}
        for c in b["chunks"]:
            c["vec"] = {int(k): v for k, v in c["vec"].items()}
            self.chunks[c["id"]] = c
        self.has_text = self.meta.get("with_text", False)

    def query_vec(self, query):
        """Query -> (qvec tf-idf, qnorm, silabas, n_oov) usando o idf da base."""
        tf, syls, oov = {}, [], 0
        for w in WORD.findall(query.lower()):
            for s in syllabify(w):
                ns = normalize(s)
                if not ns:
                    continue
                syls.append(ns)
                d = self.index.get(ns)
                if d is None:
                    oov += 1
                else:
                    tf[d] = tf.get(d, 0) + 1
        qvec, qnorm = tfidf(tf, self.idf)
        return qvec, qnorm, syls, oov

    def search(self, query, k=5, rerank=True, recall_n=20):
        qvec, qnorm, syls, oov = self.query_vec(query)
        info = {"syls": syls, "oov": oov, "dims": len(qvec),
                "n_chunks": len(self.chunks), "n_converge": 0, "recall_n": 0,
                "rerank": rerank and self.has_text, "ms_recall": 0.0, "ms_rerank": 0.0}
        if not qvec:
            return [], info
        # ESTAGIO 1 — RECALL: cosseno da query com cada chunk (vec/norm ja' na base).
        # so "convergem" os chunks que compartilham ao menos uma dimensao (cos > 0).
        t0 = time.time()
        scored = []
        for cid, c in self.chunks.items():
            s = cosine(qvec, qnorm, c["vec"], c["norm"])
            if s > 0:
                scored.append((s, cid))
        info["n_converge"] = len(scored)
        scored.sort(reverse=True)
        rn = max(k, recall_n) if info["rerank"] else k
        cand = scored[:rn]
        info["recall_n"] = min(rn, len(cand))
        info["ms_recall"] = (time.time() - t0) * 1000
        # ESTAGIO 2 — RERANK: matched filter (contiguidade) + span (proximidade)
        if info["rerank"]:
            t1 = time.time()
            res = []
            for cos, cid in cand:
                mf, span = rerank_score(query, syllable_seq(self.chunks[cid]["text"]))
                rr = mf / (1.0 + span / PROX_SCALE)
                res.append((rr, mf, span, cos, cid))
            res.sort(key=lambda t: (t[0], t[3]), reverse=True)
            hits = res[:k]
            info["ms_rerank"] = (time.time() - t1) * 1000
        else:
            hits = [(None, None, None, cos, cid) for cos, cid in cand[:k]]
        return hits, info


def show(base, query, hits, info, k):
    bar = "━" * 66
    print(f"\n{bar}\n  BUSCA: {query!r}\n{bar}")
    # [2] HISTOGRAM da query
    print(f"  [2] HISTOGRAM   query → {'-'.join(info['syls']) or '(vazio)'}")
    print(f"                  vetor da busca: {info['dims']} dims ativas, {info['oov']} OOV")
    if not hits:
        print(f"  [3] RECALL      nenhuma sílaba da query no vocabulário → 0 chunks")
        print(f"  ⇒ sem resultados.")
        return
    # [3] RECALL — quantos chunks convergem no cosseno
    print(f"  [3] RECALL      {info['n_chunks']} chunks varridos → "
          f"{info['n_converge']} convergem (cosseno > 0)   [{info['ms_recall']:.1f} ms]")
    print(f"                  ↳ mantém os top-{info['recall_n']} candidatos pro rerank")
    # [4] RERANK — filtros progressivos
    if info["rerank"]:
        print(f"  [4] RERANK      filtros: P1 contiguidade (matched filter) · "
              f"P2 proximidade (span)")
        print(f"                  ↳ {info['recall_n']} candidatos reordenados por matchpoint   "
              f"[{info['ms_rerank']:.1f} ms]")
    else:
        print(f"  [4] RERANK      (pulado — só recall por cosseno)")
    # [5] RESULTADOS — melhor (topo) → pior (base)
    print(f"  [5] RESULTADOS  top-{k}  (melhor em cima, pior embaixo):")
    for rank, (rr, mf, span, cos, cid) in enumerate(hits, 1):
        c = base.chunks[cid]
        if rr is not None:
            mp = f"matchpoint {rr:.3f}   (mf {mf:.2f} · span {span} · cos {cos:.4f})"
        else:
            mp = f"matchpoint {cos:.4f}   (cosseno puro)"
        print(f"\n   #{rank}  {mp}")
        print(f"        chunk {cid} @byte {c['start']:,}")
        if c.get("text"):
            print(f"        “{snippet(c['text'], query)}”")


def main():
    ap = argparse.ArgumentParser(
        description="Busca na base RAG (fase online): recall cosseno + rerank matched filter.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="exemplos:\n"
               "  python3 search_rag.py sda-tokenized.json \"Frodo Bolseiro\"\n"
               "  python3 search_rag.py sda-tokenized.json \"anel\" \"montanha\" -k 3\n"
               "  python3 search_rag.py sda-tokenized.json \"Gandalf\" --no-rerank\n"
               "  python3 search_rag.py sda-tokenized.json -i      # modo interativo\n")
    ap.add_argument("base", help="base JSON gerada pelo embed_gen.py (obrigatorio)")
    ap.add_argument("queries", nargs="*", help="termos de busca")
    ap.add_argument("-k", "--top", dest="k", type=int, default=5,
                    help="quantos resultados mostrar (default 5; ex: -k 10)")
    ap.add_argument("--recall-n", type=int, default=20, help="candidatos do recall p/ o rerank")
    ap.add_argument("--no-rerank", action="store_true", help="so o estagio 1 (recall cosseno)")
    ap.add_argument("-i", "--interactive", action="store_true", help="modo interativo (REPL)")
    if len(sys.argv) == 1:        # sem argumentos -> mostra o help
        ap.print_help()
        sys.exit(0)
    args = ap.parse_args()
    if not os.path.exists(args.base):
        sys.exit(f"erro: base não encontrada: {args.base!r}\n"
                 f"      gere uma com:  python3 embed_gen.py <corpus.txt>")

    t0 = time.time()
    base = RagBase(args.base)
    m = base.meta
    print(f"  [1] LOAD        base '{args.base}' em RAM   [{(time.time()-t0)*1000:.0f} ms]")
    print(f"                  {m['n_chunks']} chunks · vocab {m['vocab_size']} dims · "
          f"idf {len(base.idf)} · corpus '{m.get('corpus','?')}' (por {m.get('generator','?')})")
    if not base.has_text:
        print("                  aviso: base com --sem-texto → rerank e snippet desligados")

    rerank = not args.no_rerank
    for q in args.queries:
        hits, info = base.search(q, args.k, rerank, args.recall_n)
        show(base, q, hits, info, args.k)

    if args.interactive:
        print("\n[modo interativo — digite a busca; ENTER vazio ou 'sair' encerra]")
        while True:
            try:
                q = input("\nbusca> ").strip()
            except (EOFError, KeyboardInterrupt):
                print(); break
            if not q or q.lower() in ("sair", "quit", "exit"):
                break
            hits, info = base.search(q, args.k, rerank, args.recall_n)
            show(base, q, hits, info, args.k)


if __name__ == "__main__":
    main()

File: rust_concept/src/bin/search_rag.rs
//! search_rag (Rust) — busca na base RAG (fase online). Porte do search_rag.py.
//! recall (cosseno) + rerank (matched filter + span) + snippet, narrando os passos.
use std::collections::HashMap;
use std::io::Write;
use std::time::Instant;
use serde_json::Value;
use sylkit::{cosine, find_chars, normalize, syllabify, syllable_seq, words};

const PROX_SCALE: f64 = 8.0;

// ----------------------------- estagio 2: rerank -----------------------------
fn best_positions(qs: &[String], seq: &[String]) -> (f64, Vec<usize>) {
    let k = qs.len();
    if k == 0 || seq.len() < k { return (0.0, vec![0]); }
    let mut best: i64 = -1;
    let mut pos: Vec<usize> = vec![];
    for p in 0..=(seq.len() - k) {
        let m = (0..k).filter(|&j| seq[p + j] == qs[j]).count() as i64;
        if m > best { best = m; pos = vec![p]; }
        else if m == best { pos.push(p); }
    }
    (best as f64 / k as f64, pos)
}

fn min_span(lists: &[Vec<usize>]) -> usize {
    use std::cmp::Reverse;
    use std::collections::BinaryHeap;
    let mut heap: BinaryHeap<Reverse<(usize, usize, usize)>> = BinaryHeap::new();
    let mut cur_max = 0usize;
    for (i, lst) in lists.iter().enumerate() {
        heap.push(Reverse((lst[0], i, 0)));
        cur_max = cur_max.max(lst[0]);
    }
    let mut best = cur_max - heap.peek().unwrap().0 .0;
    loop {
        let Reverse((mn, i, j)) = heap.pop().unwrap();
        if cur_max - mn < best { best = cur_max - mn; }
        if j + 1 == lists[i].len() { return best; }
        let nxt = lists[i][j + 1];
        cur_max = cur_max.max(nxt);
        heap.push(Reverse((nxt, i, j + 1)));
    }
}

fn rerank_score(query: &str, seq: &[String]) -> (f64, usize) {
    let lower = query.to_lowercase();
    let mut words_qs: Vec<Vec<String>> = vec![];
    for w in words(&lower) {
        let qs: Vec<String> = syllabify(&w).iter().map(|s| normalize(s))
            .filter(|s| !s.is_empty()).collect();
        if !qs.is_empty() { words_qs.push(qs); }
    }
    if words_qs.is_empty() { return (0.0, 0); }
    let mut fracs = vec![];
    let mut lists = vec![];
    for qs in &words_qs {
        let (frac, pos) = best_positions(qs, seq);
        fracs.push(frac); lists.push(pos);
    }
    let mf = fracs.iter().sum::<f64>() / fracs.len() as f64;
    let span = if lists.len() > 1 { min_span(&lists) } else { 0 };
    (mf, span)
}

fn snippet(text: &str, query: &str) -> String {
    let width = 140usize;
    let flat: String = text.split_whitespace().collect::<Vec<_>>().join(" ");
    let fchars: Vec<char> = flat.chars().collect();
    let fna: Vec<char> = normalize(&flat).chars().collect();
    let n = fchars.len();
    let lowq = query.to_lowercase();
    let qwords = words(&lowq);
    let mut pos: Option<usize> = None;
    for w in &qwords {
        let wn: Vec<char> = normalize(w).chars().collect();
        if !wn.is_empty() {
            if let Some(i) = find_chars(&fna, &wn, 0) { pos = Some(i); break; }
        }
    }
    let mut out: String = match pos {
        None => {
            if n > width { let mut s: String = fchars[..width].iter().collect(); s.push('…'); s }
            else { flat.clone() }
        }
        Some(p) => {
            let a = p.saturating_sub(35);
            let b = (a + width).min(n);
            let mid: String = fchars[a..b].iter().collect::<String>().trim().to_string();
            let mut s = String::new();
            if a > 0 { s.push('…'); }
            s.push_str(&mid);
            if b < n { s.push('…'); }
            s
        }
    };
    // destaca as palavras da query com «»
    let ochars: Vec<char> = out.chars().collect();
    let ona: Vec<char> = normalize(&out).chars().collect();
    let mut marks: Vec<(usize, usize)> = vec![];
    for w in &qwords {
        let wn: Vec<char> = normalize(w).chars().collect();
        if wn.is_empty() { continue; }
        let mut start = 0;
        while let Some(i) = find_chars(&ona, &wn, start) {
            marks.push((i, i + wn.len())); start = i + wn.len();
        }
    }
    marks.sort_unstable(); marks.dedup();
    let mut res = ochars;
    for (a, b) in marks.into_iter().rev() {
        res.insert(b, '»'); res.insert(a, '«');
    }
    out = res.into_iter().collect();
    out
}

// ------------------------------- a base RAG ----------------------------------
struct Chunk { start: usize, vec: HashMap<usize, f64>, norm: f64, text: Option<String> }

struct RagBase {
    index: HashMap<String, usize>,
    idf: HashMap<usize, f64>,
    chunks: Vec<Chunk>,
    has_text: bool,
    n_chunks: usize, vocab_size: usize, corpus: String, generator: String,
}

struct Info { syls: Vec<String>, oov: usize, dims: usize, n_chunks: usize,
              n_converge: usize, recall_n: usize, rerank: bool, ms_recall: f64, ms_rerank: f64 }
type Hit = (Option<f64>, Option<f64>, Option<usize>, f64, usize); // rr, mf, span, cos, cid

impl RagBase {
    fn load(path: &str) -> RagBase {
        let data = std::fs::read_to_string(path).unwrap();
        let v: Value = serde_json::from_str(&data).unwrap();
        let meta = &v["meta"];
        let vocab: Vec<String> = meta["vocab"].as_array().unwrap().iter()
            .map(|x| x.as_str().unwrap().to_string()).collect();
        let index = vocab.iter().enumerate().map(|(i, t)| (t.clone(), i)).collect();
        let idf: HashMap<usize, f64> = v["idf"].as_object().unwrap().iter()
            .map(|(k, val)| (k.parse().unwrap(), val.as_f64().unwrap())).collect();
        let chunks: Vec<Chunk> = v["chunks"].as_array().unwrap().iter().map(|c| {
            let vec = c["vec"].as_object().unwrap().iter()
                .map(|(k, val)| (k.parse::<usize>().unwrap(), val.as_f64().unwrap())).collect();
            Chunk {
                start: c["start"].as_u64().unwrap() as usize,
                vec, norm: c["norm"].as_f64().unwrap(),
                text: c["text"].as_str().map(|s| s.to_string()),
            }
        }).collect();
        RagBase {
            index, idf,
            n_chunks: meta["n_chunks"].as_u64().unwrap() as usize,
            vocab_size: meta["vocab_size"].as_u64().unwrap() as usize,
            corpus: meta["corpus"].as_str().unwrap_or("?").to_string(),
            generator: meta["generator"].as_str().unwrap_or("?").to_string(),
            has_text: meta["with_text"].as_bool().unwrap_or(false),
            chunks,
        }
    }

    fn query_vec(&self, query: &str) -> (HashMap<usize, f64>, f64, Vec<String>, usize) {
        let lower = query.to_lowercase();
        let mut tf: HashMap<usize, u32> = HashMap::new();
        let mut syls = vec![];
        let mut oov = 0;
        for w in words(&lower) {
            for s in syllabify(&w) {
                let ns = normalize(&s);
                if ns.is_empty() { continue; }
                syls.push(ns.clone());
                match self.index.get(&ns) {
                    Some(&d) => *tf.entry(d).or_insert(0) += 1,
                    None => oov += 1,
                }
            }
        }
        let mut qvec: HashMap<usize, f64> = HashMap::new();
        for (d, c) in &tf {
            let w = *c as f64 * self.idf.get(d).copied().unwrap_or(0.0);
            if w != 0.0 { qvec.insert(*d, w); }
        }
        let s: f64 = qvec.values().map(|v| v * v).sum();
        let qnorm = if s == 0.0 { 1.0 } else { s.sqrt() };
        (qvec, qnorm, syls, oov)
    }

    fn search(&self, query: &str, k: usize, rerank: bool, recall_n: usize) -> (Vec<Hit>, Info) {
        let (qvec, qnorm, syls, oov) = self.query_vec(query);
        let mut info = Info { syls, oov, dims: qvec.len(), n_chunks: self.chunks.len(),
            n_converge: 0, recall_n: 0, rerank: rerank && self.has_text, ms_recall: 0.0, ms_rerank: 0.0 };
        if qvec.is_empty() { return (vec![], info); }
        // ESTAGIO 1 — recall por cosseno
        let t0 = Instant::now();
        let mut scored: Vec<(f64, usize)> = vec![];
        for (cid, c) in self.chunks.iter().enumerate() {
            let s = cosine(&qvec, qnorm, &c.vec, c.norm);
            if s > 0.0 { scored.push((s, cid)); }
        }
        info.n_converge = scored.len();
        scored.sort_by(|a, b| b.0.partial_cmp(&a.0).unwrap().then(b.1.cmp(&a.1)));
        let rn = if info.rerank { k.max(recall_n) } else { k };
        let cand: Vec<(f64, usize)> = scored.into_iter().take(rn).collect();
        info.recall_n = cand.len();
        info.ms_recall = t0.elapsed().as_secs_f64() * 1000.0;
        // ESTAGIO 2 — rerank
        let hits: Vec<Hit> = if info.rerank {
            let t1 = Instant::now();
            let mut res: Vec<Hit> = cand.iter().map(|&(cos, cid)| {
                let seq = syllable_seq(self.chunks[cid].text.as_deref().unwrap_or(""));
                let (mf, span) = rerank_score(query, &seq);
                let rr = mf / (1.0 + span as f64 / PROX_SCALE);
                (Some(rr), Some(mf), Some(span), cos, cid)
            }).collect();
            res.sort_by(|a, b| b.0.partial_cmp(&a.0).unwrap().then(b.3.partial_cmp(&a.3).unwrap()));
            info.ms_rerank = t1.elapsed().as_secs_f64() * 1000.0;
            res.into_iter().take(k).collect()
        } else {
            cand.into_iter().take(k).map(|(cos, cid)| (None, None, None, cos, cid)).collect()
        };
        (hits, info)
    }
}

fn commas(n: usize) -> String {
    let s = n.to_string();
    let len = s.len();
    let mut out = String::new();
    for (i, ch) in s.chars().enumerate() {
        if i > 0 && (len - i) % 3 == 0 { out.push(','); }
        out.push(ch);
    }
    out
}

fn show(base: &RagBase, query: &str, hits: &[Hit], info: &Info, k: usize) {
    let bar = "━".repeat(66);
    println!("\n{bar}\n  BUSCA: {query:?}\n{bar}");
    println!("  [2] HISTOGRAM   query → {}", if info.syls.is_empty() { "(vazio)".into() } else { info.syls.join("-") });
    println!("                  vetor da busca: {} dims ativas, {} OOV", info.dims, info.oov);
    if hits.is_empty() {
        println!("  [3] RECALL      nenhuma sílaba da query no vocabulário → 0 chunks");
        println!("  ⇒ sem resultados.");
        return;
    }
    println!("  [3] RECALL      {} chunks varridos → {} convergem (cosseno > 0)   [{:.1} ms]",
             info.n_chunks, info.n_converge, info.ms_recall);
    println!("                  ↳ mantém os top-{} candidatos pro rerank", info.recall_n);
    if info.rerank {
        println!("  [4] RERANK      filtros: P1 contiguidade (matched filter) · P2 proximidade (span)");
        println!("                  ↳ {} candidatos reordenados por matchpoint   [{:.1} ms]",
                 info.recall_n, info.ms_rerank);
    } else {
        println!("  [4] RERANK      (pulado — só recall por cosseno)");
    }
    println!("  [5] RESULTADOS  top-{k}  (melhor em cima, pior embaixo):");
    for (rank, (rr, mf, span, cos, cid)) in hits.iter().enumerate() {
        let c = &base.chunks[*cid];
        let mp = match rr {
            Some(rr) => format!("matchpoint {:.3}   (mf {:.2} · span {} · cos {:.4})",
                                rr, mf.unwrap(), span.unwrap(), cos),
            None => format!("matchpoint {:.4}   (cosseno puro)", cos),
        };
        println!("\n   #{}  {}", rank + 1, mp);
        println!("        chunk {} @byte {}", cid, commas(c.start));
        if let Some(t) = &c.text {
            println!("        “{}”", snippet(t, query));
        }
    }
}

fn help() {
    println!(
"search_rag (Rust) — busca na base RAG (recall cosseno + rerank matched filter).

uso (a base JSON e' obrigatoria — sempre explicita):
  search_rag <base.json> [queries...] [opcoes]

opcoes:
  -k, --top N      quantos resultados (default 5)
  --recall-n N     candidatos do recall p/ o rerank (default 20)
  --no-rerank      so o estagio 1 (recall cosseno)
  -i               modo interativo (REPL)

exemplos:
  search_rag sda-tokenized.json \"Frodo Bolseiro\"
  search_rag sda-tokenized.json \"anel\" \"montanha\" -k 3
  search_rag sda-tokenized.json -i");
}

fn main() {
    let argv: Vec<String> = std::env::args().collect();
    if argv.len() == 1 { help(); return; }

    let (mut k, mut recall_n) = (5usize, 20usize);
    let (mut no_rerank, mut interactive) = (false, false);
    let mut pos: Vec<String> = vec![];
    let mut it = argv[1..].iter();
    while let Some(a) = it.next() {
        match a.as_str() {
            "-h" | "--help" => { help(); return; }
            "-k" | "--top" => k = it.next().unwrap().parse().expect("-k N"),
            "--recall-n" => recall_n = it.next().unwrap().parse().expect("--recall-n N"),
            "--no-rerank" => no_rerank = true,
            "-i" | "--interactive" => interactive = true,
            other => pos.push(other.to_string()),
        }
    }
    if pos.is_empty() {
        eprintln!("erro: base JSON obrigatória.\n      uso: search_rag <base.json> [queries...]");
        std::process::exit(2);
    }
    let base_path = pos.remove(0);
    let queries = pos;
    if !std::path::Path::new(&base_path).exists() {
        eprintln!("erro: base não encontrada: {base_path:?}\n      gere uma com:  embed_gen <corpus.txt>");
        std::process::exit(1);
    }

    let t0 = Instant::now();
    let base = RagBase::load(&base_path);
    println!("  [1] LOAD        base '{}' em RAM   [{:.0} ms]", base_path, t0.elapsed().as_secs_f64() * 1000.0);
    println!("                  {} chunks · vocab {} dims · idf {} · corpus '{}' (por {})",
             commas(base.n_chunks), base.vocab_size, base.idf.len(), base.corpus, base.generator);
    if !base.has_text {
        println!("                  aviso: base com --sem-texto → rerank e snippet desligados");
    }

    let rerank = !no_rerank;
    for q in &queries {
        let (hits, info) = base.search(q, k, rerank, recall_n);
        show(&base, q, &hits, &info, k);
    }

    if interactive {
        println!("\n[modo interativo — digite a busca; ENTER vazio ou 'sair' encerra]");
        loop {
            print!("\nbusca> ");
            std::io::stdout().flush().ok();
            let mut line = String::new();
            if std::io::stdin().read_line(&mut line).unwrap_or(0) == 0 { println!(); break; }
            let q = line.trim();
            if q.is_empty() || matches!(q.to_lowercase().as_str(), "sair" | "quit" | "exit") { break; }
            let (hits, info) = base.search(q, k, rerank, recall_n);
            show(&base, q, &hits, &info, k);
        }
    }
}
